Python 3.9 syntax support

adam-azarchs opened the pull request:
Updates parser grammar and ASDL for python 3.8.

This proved a bit more invasive than I'd originally hoped because the unit tests depend on comparing output with the python interpreter, which means flowing changes through to the compiler as well.

I do not plan to make the new features actually work, only for the parser to tolerate them when building the AST.

Closes https://github.com/go-python/gpython/issues/136.

adam-azarchs committed:
Update Python.asdl and Grammar to python 3.8.

Taken from
https://github.com/python/cpython/blob/3.8/Parser/Python.asdl
https://github.com/python/cpython/blob/3.8/Parser/asdl.py
https://raw.githubusercontent.com/python/cpython/3.8/Grammar/Grammar

and make edits to asdl_go.py to make it run.

Add new AST types from python 3.8 ASDL update.

- AsyncFunctionDef
- AnnAssign (annotated assignment, e.g. `x: float = 4`)
- AsyncFor
- AsyncWith
- NamedExpr (name := expr)
- Await

Num, Str, Bytes are now all just Constant, and there's no more
NamedConstant.

Formatting fixes to make dump consistent with cpython.

Update make_grammar_test.py to exercise new grammar features.

adam-azarchs committed:
Update tests and other packages for python 3.8.

Also CI infrastructure.

Add vm opcodes.  Keep old, removed ones in place so as to not break the
build.

Fix python tests which were wrong for 3.8.

adam-azarchs committed:
Correct parsing for decorated async functions.

adam-azarchs committed:
Update number lexing for python 3.8.

Python 3.8 supports _ as a separator in numbers, e.g. 1_234_567.  It
also adds support for imaginary numbers with a j suffix.

drew-512 commented:
Hi @adam-azarchs it looks like this got left -- is this something you are interested in proceeding with?  I think this would be a favorable add if you would like to pick this up.

adam-azarchs commented:
Yeah, thanks for checking in.  Unfortunately, I don't know when I'll have time to push this through the rest of the way.  Our use case for `gpython` only requires us to be able to parse python from go; we don't actually need to have the runtime parts up and running, which makes it hard for me to justify spending a lot of time on that part.

It might be reasonable to just make sure there are no regressions, but make it clear that while the parser can build an AST for python 3.8, none of those new features are actually supported?  There's a few in there which wouldn't be easy, particularly `async`.  I also kind of kludged f-strings - the parser as implemented is more or less ignoring the `f` prefix and not worrying about what's inside (that could be fixed relatively easily, just hasn't been a priority).

glaukiol1 committed:
stdlib/os: first import

author glaukiol1 <glaukiol1@gmail.com>
committer Sebastien Binet <binet@cern.ch>

Tatskaari committed:
Merge master

Commit:
Fix ** and * args in Calls

Tatskaari commented:
Hey! I was just trying to get to grips with how this library works and decided to try and fix your tests:

Managed to fix something in the compiler here:
https://github.com/10XGenomics/gpython/commit/9c8103a92772502e7391d9cd8dba0b9a38659c9b

Seems the special opcodes for variable args require that they're the last expressions on the stack.

Commit:
all: improve error logging when parsing a file

adam-azarchs committed:
Merge 'master' into azarchs/py38

adam-azarchs committed:
Merge commit '9c8103a92772502e7391d9cd8dba0b9a38659c9b' into azarchs/py38

adam-azarchs commented:
@Tatskaari Thanks, I've merged your commit into this branch, as well as merging upstream master in.

sbinet committed:
repl: reorder want/got into got/want

This CL reverts the display of want/got into got/want in order to
follow good Go practices.
Also align got/want display so as to ease deciphering the error
for normal humans.

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
repl: make replTest.assert a T.Helper

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
pytest: store output of tested script

This CL stores the output of a failing script for easy human comparison

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
py: add convenience function Println

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
ci: add static-check test

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
ci: reduce git-fetch depth to 1

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
compile: silence staticcheck

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
all: apply gofmt

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
all: use go:build directives

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
all: apply staticcheck fixes

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
stdlib/glob: first import

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
py: improve ParseTuple{,AndKeywords} to handle 's{,*,#}'

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
py: add str.replace

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
py: improve ParseTuple{,AndKeywords} to handle 'y{,*,#}'

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
py: add bytes.replace

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
all: move marshal to stdlib/marshal

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
stdlib/binascii: first import

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
py: make Exception implement __{str,repr}__

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
stdlib/binascii: implement {b2a,a2b}_qp

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
py: make bytes implement __{i,}add__

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
all: rename master into main

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
py: add String.find

Updates #139.

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
py: handle end=-1 in str.find

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
py: handle corner cases for str.find

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
stdlib/tempfile: first import

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
stdlib/os: add mkdir, makedirs, remove, removedirs and rmdir

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
py: introduce FileModeFrom

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
stdlib/os: implement fdopen

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
stdlib/os: add close

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
stdlib/tempfile: use os.close

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
pytest: refactor testing infrastructure

This CL also exposes a -regen flag to easily regenerate golden files.

Co-authored-by: Drew O'Meara <drew@plan-systems.org>
Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
examples/embedding: use new pytest testing scaffolding

Co-authored-by: Drew O'Meara <drew@plan-systems.org>
Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
all: bump to Go-1.19

Signed-off-by: Sebastien Binet <binet@cern.ch>

sbinet committed:
all: bump go-cmp@v0.5.8

Signed-off-by: Sebastien Binet <binet@cern.ch>

drew-512 committed:
all: minor comment and typos cleanups

Co-authored-by: Drew O'Meara <drew@plan-systems.org>

adam-azarchs committed:
Fix assignment expressions in elif.

adam-azarchs committed:
Merge 'upstream/main' into azarchs/py38

adam-azarchs committed:
Update to python 3.9.

Mostly, anyway.  The Grammar file is no longer authoritative, having
been replaced with the python.gram PEG file, which I haven't switched
this to use.
